tests/handlers/test_sync.py

```diff
@@ -186,6 +187,139 @@ def test_unknown_room_version(self):
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[last_room_creation_event_id],
+                type=EventTypes.Member,
+                state_key=moderator,
+                room_id=room_id,
+                content={"membership": "join"},
+            )
+        )
+
+        # Alice makes Bob a moderator.
+        room_power_levels = self.helper.get_state(
+            room_id,
+            EventTypes.PowerLevels,
+            tok=owner_tok,
+        )
+        room_power_levels["users"][moderator] = 50
+        moderator_pl_event_id = self.helper.send_state(
+            room_id,
+            EventTypes.PowerLevels,
+            room_power_levels,
+            tok=owner_tok,
+        )["event_id"]
+
+        # We now want to concoct a situation where:
+        # - another local user Eve attempts to join the room
+        # - the homeserver generates a join event with prev_events that precede the ban
+        #   (so that it passes the "are you banned" test)
+        # - but the join event has a stream_ordering after that of the ban (so that it
+        #   comes down /sync after the ban).
+        #
+        # We do this by
+        # 1. Ban eve from the room.
+        # 2. Have eve join the room.
+        #    - mock the prev_events for the join event to precede the ban
+
+        # First, the ban.
+        eve = self.register_user("eve", "password")
+        eve_token = self.login(eve, "password")
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[moderator_pl_event_id],
+                type=EventTypes.Member,
+                state_key=eve,
+                room_id=room_id,
+                content={"membership": "ban"},
+            )
+        )
+
+        # We need a sync_token for eve after the ban.
+        eve_requester = create_requester(eve)
+        eve_sync_config = generate_sync_config(eve)
+        eve_sync_after_ban: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(eve_requester, eve_sync_config)
+        )
+        since_token = eve_sync_after_ban.next_batch
+        # Sanity check this sync result. We should be banned from the room.
+        self.assertEqual(len(eve_sync_after_ban.joined), 0)
+        self.assertEqual(len(eve_sync_after_ban.archived), 1)
+        self.assertEqual(eve_sync_after_ban.archived[0].room_id, room_id)
+
+        mocked_get_prev_events = patch.object(
+            self.hs.get_datastore(),
+            "get_prev_events_for_room",
+            new_callable=MagicMock,
+            return_value=make_awaitable([moderator_pl_event_id]),
+        )
+        with mocked_get_prev_events:
+            self.helper.join(room_id, eve, tok=eve_token)
+
+        # Eve makes a second, incremental sync. The join event should appear, but not
+        # under the "joined" section.
```

Comment:
We don't seem to be asserting that the join event appears?

Reply:
I think this comment might be wrong (it certainly wasn't the behaviour I saw). I think I might need to double check the scenario with @richvdh here.